### PR TITLE
feat: update serverless yml + ingestData function

### PR DESCRIPTION
**Pull request by ilhaoni**
[Trello](https://trello.com/c/9cncjiss)

- [x] update serverless to spin up an s3 bucket, dynamodb and lambda function per mystique environment
- [x] fetch filename and file contents when a new file is uploaded into s3
- [x] save file contents into dynamodb
- [x] look into security groups / vpc to connect lambda and s3 together, without opening up the bucket and files inside to public access :x 

How to test: 
- Look for the s3 bucket specified in the serverless.yml
- Add a new csv file, following the same format as the other files. (when you upload, set the file to public access. if its not set to public access, lambda function doesnt have permission to access the file).

Next steps to do (in other PRs):
- update proxy function handler to query from db
- decide on error to throw if there's no results from the db
- update the db schema, currently if theres a new entry, we simply overwrite the row in the db, cause now id is set to be unique key :x

### CODE REVIEW

File: src/functionHandlers/ingestData/ingestData.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import { getLogger } from "../../common/logger";
+import { put } from "../../services/db";
+import { S3EventMessage, MappingRecord } from "../../types";
+
+const { info } = getLogger("ingestData");
+
+interface CsvRow {
+  field1: string;
+  field2: string;
+  field3: number;
```
Comment (szenius) [inline src/functionHandlers/ingestData/ingestData.ts:14]: Are these going to be the header names for the CSV files?

File: src/functionHandlers/ingestData/ingestData.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+interface MappingRow {
+  TableName: string;
+  Item: MappingRecord;
+}
+
+const handleRequest = middy((event: S3EventMessage) => {
+  const objectKey = event.Records[0].s3.object.key;
+  const objectVersion = event.Records[0].s3.object.versionId;
+  const bucketName = event.Records[0].s3.bucket.name;
+  info(`${JSON.stringify({ code: "S3_FILE_CREATED", bucketName, objectKey, objectVersion })}`);
```
Comment (szenius) [inline src/functionHandlers/ingestData/ingestData.ts:26]: Nit. Would `UPLOADED`/`ADDED` be more appropriate here?

File: src/functionHandlers/ingestData/ingestData.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+interface CsvRow {
+  field1: string;
+  field2: string;
+  field3: number;
+}
+
+interface MappingRow {
+  TableName: string;
+  Item: MappingRecord;
+}
```
Comment (szenius) [inline src/functionHandlers/ingestData/ingestData.ts:20]: Hmm for this and `CsvRow`, is there a reason we don't want to put in `types.ts`?

File: src/functionHandlers/ingestData/ingestData.ts (hunk 1 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:14]: Nope nope, the csv files that we upload do not have a header haha. Hence the library defaulted to field1, 2 and 3

File: src/functionHandlers/ingestData/ingestData.ts (hunk 2 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:26]: HAHA not nit. Let me change HAHA

File: src/functionHandlers/ingestData/ingestData.ts (hunk 3 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:20]: hmm, it feels like it wouldnt be used anywhere else, so.. hence I put it here haha. My impression of types is that it'll be used across multiple files haha.

but i can see why it also belongs in types.

🙃  ermz

**Comment by ilhaoni** [comment]
> full test flow works for me!
> slight concern wrt to granting public access for s3 bucket, any way we can use a specific IAM role to grant access for the lambda function to read s3? (just thinking out loud)

yayy~ yeah i'm concerned with the public access too haha. 

i actually granted the iam access on the lambdas to access the s3 bucket and everything inside it ._. but it didnt work. or maybe we need an iam role on the s3 bucket side also? :O

File: src/functionHandlers/ingestData/ingestData.ts (hunk 3 above)
Comment (szenius) [reply src/functionHandlers/ingestData/ingestData.ts:20]: HAHA no issue with this, but I thought we were going to centralise all the interfaces and types in `types.ts`. Up to you bah!!

Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:20]: HAHA, its ok~ lets centralise HAHA

File: src/functionHandlers/ingestData/ingestData.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  info(`${JSON.stringify({ code: "S3_FILE_CREATED", bucketName, objectKey, objectVersion })}`);
+
+  const s3 = new aws.S3({
+    region: "ap-southeast-1",
+    maxRetries: 2
+  });
+
+  const params = { Bucket: bucketName, Key: objectKey };
+  const s3Stream = s3.getObject(params).createReadStream();
+  csv({ noheader: true })
```
Comment (rjchow) [inline src/functionHandlers/ingestData/ingestData.ts:35]: what happens if csv is malformed and/or not a csv?

File: src/functionHandlers/ingestData/ingestData.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    .fromStream(s3Stream)
+    .on("data", row => {
+      const jsonRow: CsvRow = JSON.parse(row);
```
Comment (rjchow) [inline src/functionHandlers/ingestData/ingestData.ts:38]: should validate the data

File: src/functionHandlers/ingestData/ingestData.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import aws from "aws-sdk";
+import { s3TriggerHandler } from "../../middlewares/handlers";
+import { getLogger } from "../../common/logger";
+import { put } from "../../services/db";
+import { S3EventMessage, MappingRecord } from "../../types";
+
+const { info } = getLogger("ingestData");
+
+interface CsvRow {
+  field1: string;
```
Comment (rjchow) [inline src/functionHandlers/ingestData/ingestData.ts:12]: document the format of these data or else someone will curse you 18 levels in the future

File: src/functionHandlers/ingestData/ingestData.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  field3: number;
+}
+
+interface MappingRow {
+  TableName: string;
+  Item: MappingRecord;
+}
+
+const handleRequest = middy((event: S3EventMessage) => {
+  const objectKey = event.Records[0].s3.object.key;
```
Comment (rjchow) [inline src/functionHandlers/ingestData/ingestData.ts:23]: what happens if Records[] has more than 1 obj? i'm not familiar with the s3 event format, might be helpful to someone working on this in the future to document the format of the event

File: src/functionHandlers/ingestData/ingestData.ts (hunk 2 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts]: fixed!

File: src/functionHandlers/ingestData/ingestData.ts (hunk 3 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts]: fixed!

File: src/functionHandlers/ingestData/ingestData.ts (hunk 5 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:27]: what do you mean by this ah

File: src/functionHandlers/ingestData/ingestData.ts (hunk 4 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:24]: so far in serverless.yml, the trigger will only work for files with suffixes .csv haha. malformed csv.. i need to test

File: src/functionHandlers/ingestData/ingestData.ts (hunk 5 above)
Comment (rjchow) [reply src/functionHandlers/ingestData/ingestData.ts:27]: check that field1,2,3 contents won't break your database

Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:28]: added!

File: src/functionHandlers/ingestData/ingestData.ts (hunk 6 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts]: i documented liao :x HAHA

File: src/functionHandlers/ingestData/ingestData.ts (hunk 7 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:13]: Hmm, so far it's set to trigger on each object upload, so i dont think there'll be more than 1 obj inside the records. I added a url to point to the full event format, would that be good enough? :x 

the interface i created is just a subset based on the fields that i need haha

File: src/functionHandlers/ingestData/ingestData.ts (hunk 4 above)
Comment (ilhaoni) [reply src/functionHandlers/ingestData/ingestData.ts:22]: so far the malformed csv will
- fail the validation
- log an error 
- do nothing

File: src/functionHandlers/ingestData/ingestData.ts (hunk 7 above)
Comment (rjchow) [reply src/functionHandlers/ingestData/ingestData.ts:13]: https://forums.aws.amazon.com/thread.jspa?messageID=592264#592264

File: src/functionHandlers/ingestData/ingestData.ts (hunk 4 above)
Comment (rjchow) [reply src/functionHandlers/ingestData/ingestData.ts:22]: as long as you will notice the error somewhere that’s fine 😆

**Pull request merged**